Site builder build promise

mbland opened the pull request:
cc: @msecret @rogeruiz

Commit:
Return a Promise from generateBuildDone callback

Commit:
generateBuildDone => finishBuild, return a Promise

Commit:
Wrap SiteBuilder.build code inside a Promise

mbland commented:
Shoulda mentioned, there's another commit in the pipe to return a `Promise` from `SiteBuilder.build` after this, but #29 and the follow-up commit to #32 for test portability need to go in first.

Commit:
Upgrade file-locked-operation to v1.0.0

FileLockedOperation is now a custom type, and
FileLockedOperation.doLockedOperation now returns a Promise.

msecret commented on lib/site-builder.js:
```diff
@@ -43,12 +43,12 @@ function SiteBuilder(opts, branch, commandRunner, jekyllHelper,
 
-SiteBuilder.prototype.generateBuildDone = function(done) {
+SiteBuilder.prototype.finishBuild = function() {
   var builder = this,
```
Sorry if I've missed this earlier.

Usual JS convention dictates to use the var name `self` for local scoping `this`. Is that possible here, or would it be too much work?

mbland replied:
It's definitely possible, and certainly not too much work. However, even though this is all server-side code, apparently [`self` has been appropriated for the global context](https://developer.mozilla.org/en-US/docs/Web/API/Window/self), so I'd be hesitant to use it that way here. However, you're the more experienced JS dev here, so if you say I'm being paranoid, I'll happily change it.

I've formerly done `var that = this` (and there are still examples of it in this PR), but I've started trying to use names that are more descriptive of the object, hence the selection of `builder`. However, like I said, if you say I should go with `self` or `that`, I'm happy to do so.

msecret commented on lib/site-builder.js:
```diff
@@ -70,18 +71,31 @@ SiteBuilder.prototype.spawn = function(path, args) {
+      });
+    })
+    .then(function() { return builder.checkForFile('_config.yml'); })
+    .then(function(useJekyll) {
```
Could this promise chain use a `catch` in case an error occurs?

mbland replied:
`builder.updateLock.doLockedOperation(doBuild)` will propagate any rejections, and the `catch` block in the return statement at the end of `SiteBuilder.build` will handle the failure case before propagating the rejection further.

msecret commented on lib/site-builder.js:
```diff
@@ -70,18 +71,31 @@ SiteBuilder.prototype.spawn = function(path, args) {
+  finish = function(err) {
+    return builder.finishBuild()
+      .then(function() {
+        done(err);
```
Is this `done()` function only called when there's an error? Is there a reason this was kept as a callback rather then being switched to a promise?

mbland replied:
Not exactly, but I've updated both this function and `finishBuild()` in a way that is hopefully more straightforward and readable.

Commit:
Refactor finishBuild(), build() return statement

Per @msecret in #33, made the `finish` function more straightforward by
passing it an `err` argument, which `finishBuild()` will then use to reject or
resolve its promise. This means that calling `done` as the final handler is
straightforward, and easily removed in a later PR when we return the resulting
`Promise` directly.
